hw/top_earlgrey/dv/env/seq_lib/chip_sw_base_vseq.sv

```
@@ -643,6 +648,42 @@ class chip_sw_base_vseq extends chip_base_vseq;
     `uvm_info(`gfn, "LC transition request succeed!", UVM_LOW)
   endtask
 
+  // Use JTAG interface to program OTP fields.
+  virtual task jtag_otp_program32(int addr,
+                                  bit [31:0] data);
+
+    bit [TL_DW-1:0] status;
+    bit idle = 0;
+    int base_addr = top_earlgrey_pkg::TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR;
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_address.get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         addr);
+
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_wdata[0].get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         data[31:0]);
+
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_cmd.get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         1 << ral.otp_ctrl_core.direct_access_cmd.wr.get_lsb_pos());
+
+
+    while (!idle) begin
+      jtag_riscv_agent_pkg::jtag_read_csr(base_addr +
+                                          ral.otp_ctrl_core.status.get_offset(),
+                                          p_sequencer.jtag_sequencer_h,
+                                          status);
+
+      idle = status >> (ral.otp_ctrl_core.status.dai_idle.get_lsb_pos()) & TL_DW'(1'b1);
+      `uvm_info(`gfn, $sformatf("Waiting for DAI to become idle = 1, actual: %d!", idle),
+         UVM_MEDIUM)
```

I guess we do not care if this operation has any error?
Or we could return the status field and let user decide if they want to use it?

---

that's a good point, we should return the error and let the test sequence decide.

---

updated! can you have a look?

---

looks great. thanks